go/test/endtoend/shardedrecovery/sharded_recovery_test.go

```diff
@@ -0,0 +1,630 @@
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard1Name), 1)
+	require.Nil(t, err)
+
+	// Build vtgate grpc connection
+	// check that vtgate doesn't route queries to new tablet
+	grpcAddress := fmt.Sprintf("%s:%d", localCluster.Hostname, localCluster.VtgateGrpcPort)
+	vtgateConn, err := vtgateconn.Dial(context.Background(), grpcAddress)
+	assert.Nil(t, err)
+	session := vtgateConn.Session("@master", nil)
+	executeQueriesUsingVtgate(t, session, "insert into vt_insert_test (id, msg) values (2,'test 2')")
+	executeQueriesUsingVtgate(t, session, "insert into vt_insert_test (id, msg) values (3,'test 3')")
+
+	vtgateConn.Close()
+	err = vtgateInstance.TearDown()
+	assert.Nil(t, err)
+
+	// now bring up the recovery keyspace and 2 tablets, letting it restore from backup.
+	restoreTablet(t, replica2, recoveryKS, "-80")
+	restoreTablet(t, replica3, recoveryKS, "80-")
+
+	// check the new replicas have the correct number of rows
+	checkData(t, replica2, shard0Count)
+	checkData(t, replica3, shard1Count)
+
+	// start vtgate
+	vtgateInstance = localCluster.GetVtgateInstance()
+	vtgateInstance.TabletTypesToWait = "REPLICA"
+	err = vtgateInstance.Setup()
+	localCluster.VtgateGrpcPort = vtgateInstance.GrpcPort
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", keyspaceName, shard1Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.master", keyspaceName, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", keyspaceName, shard1Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", recoveryKS, shard0Name), 1)
+	require.Nil(t, err)
+	err = vtgateInstance.WaitForStatusOfTabletInShard(fmt.Sprintf("%s.%s.replica", recoveryKS, shard1Name), 1)
+	require.Nil(t, err)
+
+	// check that vtgate doesn't route queries to new tablet
+	grpcAddress = fmt.Sprintf("%s:%d", localCluster.Hostname, localCluster.VtgateGrpcPort)
+	vtgateConn, err = vtgateconn.Dial(context.Background(), grpcAddress)
+	assert.Nil(t, err)
+	session = vtgateConn.Session("@replica", nil)
+	verifyQueriesUsingVtgate(t, session, "select count(*) from vt_insert_test", "INT64(4)")
+
+	// check that new keyspace is accessible by using ks.table
+	verifyQueriesUsingVtgate(t, session, "select count(*) from recovery_keyspace.vt_insert_test", "INT64(2)")
+
+	// check that new keyspace is accessible with 'use ks'
+	executeQueriesUsingVtgate(t, session, "use recovery_keyspace@replica")
+	verifyQueriesUsingVtgate(t, session, "select count(*) from vt_insert_test", "INT64(2)")
+
```

Comment:
Same here, we should add testing for keyspace:shard@replica
https://github.com/vitessio/vitess/blob/master/test/sharded_recovery.py#L611

Reply:
done